File: api/Helpers/Result.cs
namespace ProjectManagementAPI.Helpers;

// ═══════════════════════════════════════════════════════════════════════════
//  Result<T> — สำหรับ failure ที่ "คาดไว้แล้ว"
//
//  "page not found", "rank collision", "formula cycle detected" คือ outcome
//  ไม่ใช่ exception exception เก็บไว้ใช้กับบั๊กจริง ๆ เท่านั้น
//
//  เหตุผลที่ไม่ใช้ exception เป็น control flow:
//    1. อ่าน signature แล้วรู้เลยว่า method นี้ล้มเหลวได้แบบไหน
//    2. DocHub.PushUpdate ทำงานทุก keystroke — throw/catch ใน hot path
//       เป็นปัญหา performance จริง ไม่ใช่แค่เรื่องสไตล์
// ═══════════════════════════════════════════════════════════════════════════

/// <summary>ชนิดของความล้มเหลว — controller แปลงเป็น HTTP status ที่เดียว</summary>
public enum ErrorKind
{
    /// <summary>ไม่พบ — หรือมีแต่ผู้ใช้ไม่มีสิทธิ์เห็น (ตอบ 404 ทั้งสองกรณี)</summary>
    NotFound,

    /// <summary>input ไม่ถูกต้อง → 400</summary>
    Validation,

    /// <summary>ยังไม่ได้ login → 401</summary>
    Unauthorized,

    /// <summary>login แล้วแต่ไม่มีสิทธิ์ → 403</summary>
    Forbidden,

    /// <summary>ชนกับ state ปัจจุบัน เช่น slug ซ้ำ → 409</summary>
    Conflict,

    /// <summary>ระบบภายนอกล้ม → 503</summary>
    Unavailable
}

public readonly record struct Error(ErrorKind Kind, string Message, string? Code = null)
{
    public static Error NotFound(string message, string? code = null)
        => new(ErrorKind.NotFound, message, code);

    public static Error Validation(string message, string? code = null)
        => new(ErrorKind.Validation, message, code);

    public static Error Unauthorized(string message, string? code = null)
        => new(ErrorKind.Unauthorized, message, code);

    public static Error Forbidden(string message, string? code = null)
        => new(ErrorKind.Forbidden, message, code);

    public static Error Conflict(string message, string? code = null)
        => new(ErrorKind.Conflict, message, code);

    public static Error Unavailable(string message, string? code = null)
        => new(ErrorKind.Unavailable, message, code);
}

/// <summary>ผลลัพธ์ที่มีค่ากลับ</summary>
public readonly struct Result<T>
{
    private readonly T? _value;

    private Result(T value)
    {
        _value = value;
        Error = default;
        IsSuccess = true;
    }

    private Result(Error error)
    {
        _value = default;
        Error = error;
        IsSuccess = false;
    }

    public bool IsSuccess { get; }
    public bool IsFailure => !IsSuccess;
    public Error Error { get; }

    /// <summary>อ่านค่าได้เมื่อ IsSuccess เท่านั้น</summary>
    public T Value => IsSuccess
        ? _value!
        : throw new InvalidOperationException(
            $"อ่าน Value จาก Result ที่ล้มเหลวไม่ได้ ({Error.Kind}: {Error.Message})");

    public static Result<T> Success(T value) => new(value);
    public static Result<T> Failure(Error error) => new(error);

    // ทำให้เขียน `return someValue;` และ `return Error.NotFound(...)` ได้ตรง ๆ
    public static implicit operator Result<T>(T value) => new(value);
    public static implicit operator Result<T>(Error error) => new(error);
}

/// <summary>ผลลัพธ์ที่ไม่มีค่ากลับ (command)</summary>
public readonly struct Result
{
    private Result(Error error)
    {
        Error = error;
        IsSuccess = false;
    }

    private Result(bool success)
    {
        Error = default;
        IsSuccess = success;
    }

    public bool IsSuccess { get; }
    public bool IsFailure => !IsSuccess;
    public Error Error { get; }

    public static Result Success() => new(true);
    public static Result Failure(Error error) => new(error);

    public static implicit operator Result(Error error) => new(error);
}

File: server/src/common/result.ts
// ═══════════════════════════════════════════════════════════════════════════
//  Result<T> — สำหรับ failure ที่ "คาดไว้แล้ว"
//
//  "ไม่พบหน้า", "slug ซ้ำ", "รหัสผ่านผิด" คือ outcome ไม่ใช่ exception
//  exception เก็บไว้ใช้กับบั๊กจริง ๆ เท่านั้น
//
//  ── ต่างจากฝั่ง C# ตรงนี้ ────────────────────────────────────────────────
//  ของเดิมให้เหตุผลไว้สองข้อ ข้อแรกยังจริงทั้งหมด ข้อสองไม่จริงแล้ว:
//
//    1. อ่าน signature แล้วรู้เลยว่าฟังก์ชันนี้ล้มเหลวได้แบบไหน  ← ยังใช่
//    2. throw/catch ใน hot path (ทุก keystroke) แพง                ← ไม่ใช่แล้ว
//
//  ข้อสองหายไปเพราะ hot path ของ Yjs ย้ายไปอยู่บน WebSocket ไม่ผ่าน HTTP
//  แล้ว การ throw จึงเกิดแค่ครั้งเดียวต่อ request ที่ล้มเหลว ที่ขอบ controller
//
//  ผลคือ service ยังคืน Result (ได้ข้อ 1 มาเต็ม ๆ) แต่ controller เรียก
//  unwrap() ซึ่ง throw HttpException ให้ Nest จัดการต่อตามปกติ — ไม่ต้องมี
//  interceptor พิเศษมาแกะ และ Swagger/filter ทำงานเหมือน endpoint ทั่วไป
// ═══════════════════════════════════════════════════════════════════════════

/** ชนิดของความล้มเหลว — แปลงเป็น HTTP status ที่เดียวใน api-response.ts */
export type ErrorKind =
  /** ไม่พบ — หรือมีแต่ผู้ใช้ไม่มีสิทธิ์เห็น (ตอบ 404 ทั้งสองกรณี) */
  | 'not_found'
  /** input ไม่ถูกต้อง → 400 */
  | 'validation'
  /** ยังไม่ได้ login → 401 */
  | 'unauthorized'
  /** login แล้วแต่ไม่มีสิทธิ์ → 403 */
  | 'forbidden'
  /** ชนกับ state ปัจจุบัน เช่น slug ซ้ำ → 409 */
  | 'conflict'
  /** ระบบภายนอกล้ม → 503 */
  | 'unavailable';

export interface AppError {
  readonly kind: ErrorKind;
  readonly message: string;
  readonly code?: string;
}

/**
 * ⚠️ ฝั่งล้มเหลวเป็น type ของตัวเอง ไม่ใช่แค่ member ของ union
 *
 *    ทำให้ `err.notFound(...)` มี `.error` ให้อ่านได้ทันทีโดยไม่ต้องแคบชนิดก่อน
 *    (interceptor ต้องใช้เพื่อสร้าง ApiException) และยัง assign เข้า Result<T>
 *    ของ T อะไรก็ได้ตามปกติ
 */
export interface Failure {
  readonly ok: false;
  readonly error: AppError;
}

export type Success<T> = { readonly ok: true; readonly value: T };
export type Result<T> = Success<T> | Failure;

/** command ที่ไม่มีค่ากลับ */
export type VoidResult = Result<null>;

export const ok = <T>(value: T): Result<T> => ({ ok: true, value });
export const okVoid = (): VoidResult => ({ ok: true, value: null });

const fail = (kind: ErrorKind) => (message: string, code?: string): Failure => ({
  ok: false,
  error: code === undefined ? { kind, message } : { kind, message, code },
});

export const err = {
  notFound: fail('not_found'),
  validation: fail('validation'),
  unauthorized: fail('unauthorized'),
  forbidden: fail('forbidden'),
  conflict: fail('conflict'),
  unavailable: fail('unavailable'),
};

/**
 * ส่งต่อความล้มเหลวจาก Result ตัวอื่นโดยไม่ต้องแกะ error ออกมาเอง
 *
 * TypeScript แคบชนิดให้เองหลัง `if (!r.ok) return propagate(r)` จึงไม่ต้อง
 * cast ที่จุดเรียก
 */
export const propagate = (result: Failure): Failure => result;
